Plutarch/Builtin/List.hs

```diff
@@ -0,0 +1,111 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Plutarch.Builtin.List (
+  -- * List type
+  PList (..),
+
+  -- * List construction
+  cons,
+  nil,
+  mkList,
+  singleton,
+  append,
+
+  -- * List access
+  head,
+  contains,
+  atIndex,
+) where
+
+import Data.Proxy
+import Plutarch
+import Plutarch.Bool
+import Plutarch.Builtin
+import Plutarch.Builtin.List.Type
+import Plutarch.Integer (PInteger)
+import Plutarch.Prelude
+import qualified PlutusCore as PLC
+import Prelude hiding (head)
+
+instance ListElemUni (a :: k -> Type) => PlutusType (PList a) where
+  type PInner (PList a) _ = PList a
+  pcon' PNil =
+    punsafeConstant $
+      PLC.Some $
+        PLC.ValueOf (PLC.DefaultUniList $ listElemUni (Proxy :: Proxy a)) []
+  pcon' (PCons x xs) = MkCons #£ x £ xs
+  pmatch' = pmatchList
+
+pmatchList :: forall k (s :: k) (a :: k -> Type) (b :: k -> Type). Term s (PList a) -> (PList a s -> Term s b) -> Term s b
+pmatchList list f =
+  plet (NullList #£ list) $ \isEmpty ->
+    pif
+      (punsafeCoerce isEmpty)
+      (f PNil)
+      $ plet
+        (HeadList #£ list)
+        ( \head ->
+            plet (TailList #£ list) $ \tail ->
+              f $ PCons head tail
+        )
+
+cons :: forall k (s :: k) (a :: k -> Type). ListElemUni a => Term s a -> Term s (PList a) -> Term s (PList a)
```

Comment:
Please follow the naming convention.

Reply:
The type is in addition also incorrect AFAICT.

https://github.com/input-output-hk/plutus/blob/master/plutus-metatheory/src/Algorithmic.lagda#L154

Reply:
I've since this found out that the signature in Agda is incorrect. I'll probably make an issue upstream.